crates/re_query_cache/src/flat_vec_deque.rs

```diff
@@ -0,0 +1,789 @@
+            &[8, 9, 10],
+        ],
+        &v,
+    );
+
+    v.insert_range(v.num_entries(), [vec![100], vec![200, 300, 400]]);
+    assert_deque_eq(
+        &[
+            &[20],
+            &[21],
+            &[22],
+            &[1, 2, 3],
+            &[41, 42],
+            &[43],
+            &[4, 5, 6, 7],
+            &[8, 9, 10],
+            &[100],
+            &[200, 300, 400],
+        ],
+        &v,
+    );
+}
+
+#[test]
+fn insert_with() {
+    let mut v: FlatVecDeque<i64> = FlatVecDeque::new();
+
+    assert_eq!(0, v.num_entries());
+    assert_eq!(0, v.num_values());
+
+    v.insert_with(
+        0,
+        FlatVecDeque::from_vecs([vec![1, 2, 3], vec![4, 5, 6, 7], vec![8, 9, 10]]),
+    );
+    assert_deque_eq(&[&[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]], &v);
+
+    v.insert_with(0, FlatVecDeque::from_vecs([vec![20], vec![21], vec![22]]));
+    assert_deque_eq(
+        &[&[20], &[21], &[22], &[1, 2, 3], &[4, 5, 6, 7], &[8, 9, 10]],
+        &v,
+    );
+
+    v.insert_with(4, FlatVecDeque::from_vecs([vec![41, 42], vec![43]]));
+    assert_deque_eq(
+        &[
+            &[20],
+            &[21],
+            &[22],
+            &[1, 2, 3],
+            &[41, 42],
+            &[43],
+            &[4, 5, 6, 7],
+            &[8, 9, 10],
+        ],
+        &v,
+    );
+
+    v.insert_with(
+        v.num_entries(),
+        FlatVecDeque::from_vecs([vec![100], vec![200, 300, 400]]),
+    );
+    assert_deque_eq(
+        &[
+            &[20],
+            &[21],
+            &[22],
+            &[1, 2, 3],
+            &[41, 42],
+            &[43],
+            &[4, 5, 6, 7],
+            &[8, 9, 10],
+            &[100],
+            &[200, 300, 400],
+        ],
+        &v,
+    );
+}
+
+// ---
+
+impl<T> FlatVecDeque<T> {
+    /// Splits the deque into two at the given index.
+    ///
+    /// Returns a newly allocated `FlatVecDeque`. `self` contains entries `[0, entry_index)`,
+    /// and the returned deque contains entries `[entry_index, num_entries)`.
+    ///
+    /// Note that the capacity of `self` does not change.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    #[must_use = "use `.truncate()` if you don't need the other half"]
+    pub fn split_off(&mut self, entry_index: usize) -> Self {
+        let value_offset = self.value_offset(entry_index);
+
+        let mut offsets = self.offsets.split_off(entry_index);
+        for offset in &mut offsets {
+            *offset -= value_offset;
+        }
+
+        Self {
+            values: self.values.split_off(value_offset),
+            offsets,
+        }
+    }
+
+    /// Shortens the deque, keeping all entries up to `entry_index` (excluded), and
+    /// dropping the rest.
+    ///
+    /// Panics if `entry_index` is out of bounds.
+    #[inline]
+    pub fn truncate(&mut self, entry_index: usize) {
+        self.offsets.truncate(entry_index);
+        self.values.truncate(self.value_offset(entry_index));
```

Comment:
It took me a bit of digging to convince myself there is no bug here, in particular that `value_offset` won't try to read off the end of the newly truncated `self.offsets`. Switching the order of the two lines makes it more obvious:

```suggestion
        self.values.truncate(self.value_offset(entry_index));
        self.offsets.truncate(entry_index);
```